library/src/google-maps-api-loader.ts

```diff
@@ -0,0 +1,215 @@
+      serializedParams === this.serializedParams &&
+      this.loadPromise
+    ) {
+      console.debug('api-loader: loading already started');
+      return this.loadPromise;
+    }
+
+    // if parameters did change, and we already loaded the API, we need
+    // to unload it first.
+    if (this.loadPromise) {
+      // FIXME: in this case, we might want to report an error if we're not
+      //  already unloading, since that would only be the case if the loader
+      //  is taked with loading multiple instances of the API with different
+      //  parameters.
+
+      // if (this.loadingState >= LoadingState.LOADING) {
+      //   console.error(
+      //     'The Google Maps API Parameters passed to the `GoogleMapsProvider` ' +
+      //       'components do not match. The Google Maps API can only be loaded ' +
+      //       'once. Please make sure to pass the same API parameters to all ' +
+      //       'of your `GoogleMapsProvider` components.'
+      //   );
+      // }
+      console.debug(
+        'api-loader: was already loaded with other params, unload first'
+      );
+      await this.unloadAsync();
+    }
+
+    console.debug('api-loader: queue request');
+
+    this.loadingState = LoadingState.QUEUED;
+    this.loadPromise = new Promise(async (resolve, reject) => {
+      console.debug('api-loader: defer to microtasks');
+      // this causes the rest of the function to be pushed back into the
+      // microtask-queue, allowing multiple synchronous calls before actually
+      // loading anything.
+      await Promise.resolve();
+      console.debug('api-loader: continue');
+
+      // if the load request was canceled in the meantime, we stop here and
+      // reject the promise. This is typically the case in react dev mode when
+      // load/unload are called from a hook.
+      if (this.loadingState !== LoadingState.QUEUED) {
+        console.debug('api-loader: no longer queued');
+        reject(new Error('map loading canceled'));
+        return;
+      }
+
+      console.debug('api-loader: start actually loading');
+      this.loadingState = LoadingState.LOADING;
+      const url = this.getApiUrl(this.params);
+
+      // setup the callback
+      window.__gmcb__ = () => {
+        console.debug(`api-loader: callback called, loading complete`);
+        this.loadingState = LoadingState.LOADED;
+        resolve();
+      };
+      url.searchParams.set('callback', '__gmcb__');
+      console.debug(`api-loader: URL: ${url}`);
+
+      // create the script
+      const script = document.createElement('script');
+      script.src = url.toString();
+      script.onerror = err => reject(err);
+
+      document.head.appendChild(script);
+    });
+
+    return this.loadPromise;
+  }
+
+  /**
+   * Unloads the Google Maps API by canceling any pending script downloads
+   * and removing the global `google.maps` object.
+   */
+  static unload(): void {
+    void this.unloadAsync();
```

Comment:
note: this is why I had to disable the no-void eslint rule. Using `void` here makes it very explicit that it wasn't just accidentally forgotten to return the promise.